benchmark/drm/publish.js

```diff
@@ -39,16 +46,38 @@ module.exports.run = function() {
         'info' : '',
         'item' : buf
     };
-    return bc.invokeSmartContract(contx, 'drm', 'v0', {verb : 'publish', item: JSON.stringify(item)}, 120)
-        .then((results)=>{
-            for (let i in results){
-                let stat = results[i];
-                if(stat.IsCommitted()) {
-                    ids.push(stat.GetResult().toString());
+
+    if (bc.bcType === 'fabric-ccp') {
```

Comment:
Benchmark test modules should give transparent support to kinds of blockchain(different types or versions), so here it would be better to reduce the tests module's dependency with the certain type (or certain version) of blockchain system. Could you please move this Fabric-ccp dependency into the file [blockchain.js](https://github.com/hyperledger/caliper/blob/master/src/comm/blockchain.js) ? And for the parameter 'invoker', is it possible to extract a common invoker instance ? Then it could be used into kinds of blockchain system.  As for the parameter 'channel', can we  add one more parameter (eg. named as 'optionalPara') with data type Map(eg. `{fabric: {channel: 'mychannel'}}` ) to the function invokeSmartContract?  What do you think about this?

Reply:
@feihujiang I thought a lot about this, and there is also TODO in the development doc about this issue (also highlighted in the [doc of the adapter](https://aklenik.github.io/caliper/docs/Fabric_Ccp_Configuration.html#the-run-callback)).

The problem is, that the abstract API of the adapters has an inconsistent signature compared to what is needed. My suggestion in the dev docs is that we adopt the Fabric SDK convention, meaning the following: the adapter functions (mainly invoke and query) take some `context` for the test round like in the current version; an `options` describing everything related to the call, and a `timeout`. 

The `options` is a platform-specific object since, for example, each platform requires a different way to identify a target smart contract (Fabric has a `channel->chaincode->version` "target ID", Sawtooth and Burrow has different ways). So the user callback can't really be platform independent, there will be a need for a "business logic -> platform syntax" translation if the user plans to target multiple platforms. 

The `if (bc.bcType === 'fabric-ccp')` line could be omitted by making the abstract adapter API consistent across platforms. But I like the idea of extracting some common concepts (like invoker) into the adapter interface. But we need to gather the common concepts of platforms first.

I'm not sure I understand the need for `{fabric: {channel: 'mychannel'}}`, could you give an example or use case for this parameter?

Reply:
No, my suggestion isn't providing the abstract API of the adapters. I mean it's better that the specific benchmark test (simple/drm etc) could be transparent to the blockchain systems.
Yes, "business logic -> platform syntax" is necessarily needed. To each platform requires a different way to identify a target smart contract, Fabric has a `channel->chaincode->version` "target ID", Sawtooth and Burrow has different ways. Could we add a mapping from a common business logic name to the target smart contract on the network Configuration File? So that the benchmark test (simple/drm etc) could use the common business logic name. If we could, {fabric: {channel: 'mychannel'}} is meaningless, avoid it.

Reply:
@feihujiang That's exactly what I was thinking after I wrote my previous comment. Just like users are identified with a unique name in this adapter, a chaincode specification inside a channel could also be labeled with a `smartContractId` attribute, that must be unique on a network level. 

So the user callback could reference it as `targetContract: simple`, and the adapter would manage the `simple => mychannel/simple/v0` mapping (similar for the other adapters). The target nodes and the invoker identity could also be referenced by name; these are probably common concepts for each (private/permissioned) platform.

However, some attributes cannot be generalized, like the target orderer to use (this is optional, but the user should be able to define it if needed), the function to invoke inside the chaincode (okay, maybe this could be common), and the syntax of transaction arguments (array of strings, named parameters, transaction dependencies, batches, etc).

So a minimal platform-specific code/arguments cannot be avoided for a multi-target benchmark. But I think it's a good tradeoff, plus this code should be really short.

I could incorporate this contract mapping into this adapter, but what about the current signature of the `invokeSmartContract` function? The separate `contractID` and `contractVersion` parameters still bother me a bit :) The new ID name should also incorporate the version because it's easy to mess up the versions in the callback (`"1.0"` vs `"v1"`)

Reply:
@aklenik Great job! It is a good way that the adapter would manage the `simple => mychannel/simple/v0` mapping (similar for the other adapters).
I agree with you that a minimal platform-specific code/arguments cannot be avoided for a multi-target benchmark. And I think trying our best to extract common concepts is benefit for users. 
The new ID name should also incorporate the version==>exactly.

Reply:
@feihujiang Then I'll add these features to this adapter, as a best-practice to follow for other adapters. So, for now, this adapter will follow the current API, but yeah, we'll need to consolidate it later 👍 
I'll ping you when it's ready.

Reply:
:+1:  :)